Stat calc

R0flcopt3r opened the pull request:
JobBare and Job now has a JobStat member with easily reachable statistics for the job.

Closes #54

R0flcopt3r committed:
docs: fix typo in docstring

R0flcopt3r committed:
feat: job stat property

tomrtk commented on src/core/api/api.py:
```diff
@@ -81,16 +81,16 @@ def convert_to_projectbare(project: model.Project) -> schema.ProjectBare:
 
 
 def convert_to_jobbare(job: model.Job) -> schema.JobBare:
```
Move this function to `utils.pt`?

R0flcopt3r replied:
It's very API specific, so not sure if it would serve any purpose to have it in `utils.py`. Would have to move `convert_to_projectbare()` too then.

MindTooth replied:
But having a `utils.py` file inside the `api` folder makes it still API specific, doesn't it?

Now it's a bit annoying that you need to scroll past these functions to actually see the routes (paths).  I know I put them there though. 😛

R0flcopt3r replied:
Fine if I resolve this, or do you still want them moved?

R0flcopt3r replied:
> But having a `utils.py` file inside the `api` folder makes it still API specific, doesn't it?
> 
> Now it's a bit annoying that you need to scroll past these functions to actually see the routes (paths). I know I put them there though. stuck_out_tongue

somehow this arrived after my comment :thinking: 

as it stands now `utils.py` is not under `api/`. I could create a `api/utils.py` and move them there if that's what you mean @MindTooth?

MindTooth replied:
Yes.  I believe that's what @tomrtk meant too.  At least we've spoken about it before.

tomrtk replied:
Yes, but we can move them later in its own PR to a separate file.

R0flcopt3r replied:
to late! done here, 350bb9c2ae1cb1f61f455bc78a47c21b4e5bdf8d

tomrtk commented on src/core/api/schema.py:
```diff
@@ -1,15 +1,18 @@
+from typing import Any, Dict, List, Optional
 
 from pydantic import BaseModel, validator
+from sqlalchemy.sql.coercions import LabeledColumnExprImpl
```
Is this in use?

tomrtk commented on src/core/api/schema.py:
```diff
@@ -100,6 +103,31 @@ class JobBase(HashableBaseModel):
+    total_labels: int
+    labels: Dict[str, int] = dict()
+
+    @validator("labels", pre=False)
```
Did you test `each_item` or `whole` on validator? I see some bug report on this in the past.
```suggestion
    @validator("labels", pre=False, each_item=True)
```

R0flcopt3r replied:
According to the docs, that's needed when you aren't iterating over the list you are validating. Ref: https://pydantic-docs.helpmanual.io/usage/validators/#subclass-validators-and-each_item

Since I am iterating over the data structure, it shouldn't be needed.

R0flcopt3r replied:
I also can't find any docs for `whole`?

tomrtk replied:
Maybe whole has been changed, but it is mentioned here when validators is running twice:

https://github.com/samuelcolvin/pydantic/issues/737
https://github.com/samuelcolvin/pydantic/issues/254

R0flcopt3r replied:
I dug around some more. `whole=True` is literally `each_item=False` (looked at the code). Which is default behavior, and does not fix the issue...

`each_item=True` is only providing the values, so I can't use that as a workaround either.

This also only happens when running through FastAPI. Simply creating a `JobStat` with the default constructor or `from_orm()` is working as expected.

I also think this is a result of us perhaps miss using the validators, using them to *translate* rather than *validating*. The current code is technically translating and validating, so it's technically okay, though somewhat messy.

tomrtk commented on tests/unit/core/integration/test_schema.py:
```diff
@@ -0,0 +1,112 @@
+    print(type(schema_stats.labels[schema.get_label(1)]))
+    assert schema_stats.labels[schema.get_label(1)] == 1
+
+    # assert schema_stats.labels == {
```
Remove this or is it to be used?

tomrtk commented on tests/unit/core/unit/test_job.py:
```diff
@@ -1,10 +1,14 @@
 from pathlib import Path
+from core import model
+
+import time
```
Is `time` used?

Move import of `model` to below `pytest`

R0flcopt3r commented:
@MindTooth why did you mark this as a draft?

MindTooth commented:
> @MindTooth why did you mark this as a draft?

Bunch of commits.

R0flcopt3r commented:
> > @MindTooth why did you mark this as a draft?
> 
> Bunch of commits.

They will all be squashed when merging.

R0flcopt3r committed:
refactor: moves api helper functions to api/utils

R0flcopt3r committed:
feat: `Job` and `JobBare` gets `JobStat` member

JobStats is the pydantic version for model.Job.stats.

Also converts labels from `int` to `str` for `JobStat`. A bug surfaced
from this where `convert_labels` happens twice. We simply ignore and
return if every key is a valid `str` that cannot be cast to an int.

R0flcopt3r committed:
test: all tests for new job statistic generation

Tests model.Job.stats property, as well as the (de)serialization with
pydantic.
